File: src/api/scheduler.ts
import { tauriInvoke } from "@api/tauri";

export interface ScheduledTask {
  id: string;
  server_id: string;
  name: string;
  enabled: boolean;
  task_type: TaskType;
  schedule: TaskSchedule;
  created_at: number;
  last_run_at: number | null;
  last_run_result: string | null;
}

export type TaskType =
  | "RestartServer"
  | { RunBackup: { incremental: boolean; include_paths: string[] } }
  | { ExecuteCommand: { command: string } };

export type TaskSchedule =
  | { Interval: { seconds: number } }
  | { Daily: { hour: number; minute: number } };

export interface TaskExecutionLog {
  id: string;
  task_id: string;
  executed_at: number;
  result: string;
  duration_ms: number;
}

export const schedulerApi = {
  async createTask(
    serverId: string,
    name: string,
    taskType: TaskType,
    schedule: TaskSchedule,
  ): Promise<ScheduledTask> {
    return tauriInvoke("create_scheduled_task", {
      serverId,
      name,
      taskType,
      schedule,
    });
  },

  async updateTask(task: ScheduledTask): Promise<ScheduledTask> {
    return tauriInvoke("update_scheduled_task", { task });
  },

  async deleteTask(taskId: string): Promise<void> {
    return tauriInvoke("delete_scheduled_task", { taskId });
  },

  async listTasks(serverId: string): Promise<ScheduledTask[]> {
    return tauriInvoke("list_scheduled_tasks", { serverId });
  },

  async toggleTask(taskId: string, enabled: boolean): Promise<void> {
    return tauriInvoke("toggle_scheduled_task", { taskId, enabled });
  },

  async getExecutionLogs(taskId: string): Promise<TaskExecutionLog[]> {
    return tauriInvoke("get_task_execution_logs", { taskId });
  },

  async runTaskNow(taskId: string): Promise<void> {
    return tauriInvoke("run_task_now", { taskId });
  },
};

File: src-tauri/src/models/scheduler.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TaskType {
    RestartServer,
    RunBackup {
        incremental: bool,
        include_paths: Vec<String>,
    },
    ExecuteCommand {
        command: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TaskSchedule {
    Interval { seconds: u64 },
    Daily { hour: u8, minute: u8 },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScheduledTask {
    pub id: String,
    pub server_id: String,
    pub name: String,
    pub enabled: bool,
    pub task_type: TaskType,
    pub schedule: TaskSchedule,
    pub created_at: u64,
    pub last_run_at: Option<u64>,
    pub last_run_result: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateTaskRequest {
    pub server_id: String,
    pub name: String,
    pub task_type: TaskType,
    pub schedule: TaskSchedule,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskExecutionLog {
    pub id: String,
    pub task_id: String,
    pub executed_at: u64,
    pub result: String,
    pub duration_ms: u64,
}
